### PR TITLE
commit order must consider non-nullability of join columns as prioritised over nullable join columns: Take 2

### PR DESCRIPTION
**Pull request by bendavies**
This is a resubmission of https://github.com/doctrine/doctrine2/pull/6533, which attempted to fix https://github.com/doctrine/doctrine2/issues/6499 (also https://github.com/doctrine/doctrine2/issues/7006)

Note that that I consulted Marco on slack before doing this.

I don't want to step on anyones toes, but the original PR has seemed to have gone stale and come to a halt. Maybe the original submitter has lost interest/moved on (his fork appears to be gone) or whatever (not judging), but regardless, I'd like to pick this back again and try and get it over the line.

I've maintained the original commits from the PR as well as some of Marco's styling fixes.

#6533 was merged and then reverted (https://github.com/doctrine/doctrine2/pull/6533#issuecomment-321911821). 
The Tests reveal the problem, `CascadeRemoveOrderTest`. If anyone has any idea how this can be fixed before I start poking around, please give me a heads up.

### CODE REVIEW

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 1]
```diff
@@ -1256,7 +1256,11 @@ private function getCommitOrder(array $entityChangeSet = null)
 
                 $joinColumns = reset($assoc['joinColumns']);
 
-                $calc->addDependency($targetClass->name, $class->name, (int)empty($joinColumns['nullable']));
+                $calc->addDependency(
+                    $targetClass->name,
+                    $class->name,
+                    (int) (($joinColumns['nullable'] ?? true) === false)
```
Comment (Ocramius) [inline lib/Doctrine/ORM/UnitOfWork.php]: I think I wrote this ternary, and honestly I should be ashamed :-\

This should be split.

**Review by Ocramius — CHANGES_REQUESTED** [review]

**Comment by Ocramius** [comment]
Restarted zeh build

File: lib/Doctrine/ORM/UnitOfWork.php (hunk 1 above)
Comment (bendavies) [reply lib/Doctrine/ORM/UnitOfWork.php]: afraid so ;).

I can split it out, but feel free to push to my branch if you have a preferred style.

Comment (bendavies) [reply lib/Doctrine/ORM/UnitOfWork.php]: done

**Comment by bendavies** [comment]
taking a look at the commit order of the entities in `CascadeRemoveOrderTest`, they are reversed with this patch on the `remove -> flush`.

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6499Test.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
+<?php
+
+namespace Doctrine\Tests\Functional\Ticket;
+
+use Doctrine\Common\Collections\ArrayCollection;
+use Doctrine\Tests\OrmFunctionalTestCase;
+
+/**
+ * @group #6499
```
Comment (lcobucci) [inline tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6499Test.php]: It should be `@group 6499` instead

**Review by lcobucci — CHANGES_REQUESTED** [review]
Just some small issues, which won't really solve the mystery...

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 3]
```diff
@@ -1255,8 +1255,14 @@ private function getCommitOrder(array $entityChangeSet = null)
...
 
                 $joinColumns = reset($assoc['joinColumns']);
+                $joinColumnsNullable = $joinColumns['nullable'] ?? true;
+                $joinColumnsNotNullable = $joinColumnsNullable === false;
 
-                $calc->addDependency($targetClass->name, $class->name, (int)empty($joinColumns['nullable']));
+                $calc->addDependency(
+                    $targetClass->name,
+                    $class->name,
+                    (int) $joinColumnsNotNullable
```
[the remark this answers is not in the document]
Comment (bendavies) [reply lib/Doctrine/ORM/UnitOfWork.php]: yeah, that seems fine.

File: lib/Doctrine/ORM/UnitOfWork.php [hunk 4]
```diff
@@ -1255,8 +1255,14 @@ private function getCommitOrder(array $entityChangeSet = null)
                 }
 
                 $joinColumns = reset($assoc['joinColumns']);
+                $joinColumnsNullable = $joinColumns['nullable'] ?? true;
```
[the remark this answers is not in the document]
Comment (bendavies) [reply lib/Doctrine/ORM/UnitOfWork.php]: this is @Ocramius code btw. I just reformatted.

[the remark this answers is not in the document]
Comment (bendavies) [reply lib/Doctrine/ORM/UnitOfWork.php]: sorry, actually the original thinking is in the description of https://github.com/doctrine/doctrine2/pull/6533

[the remark this answers is not in the document]
Comment (bendavies) [reply lib/Doctrine/ORM/UnitOfWork.php]: Taking a look, The patch introduces a difference to the weight for when `nullable` is not defined. old weight would be 1, new is 0.

```
<?php

function original(array $a): int
{
    return (int)empty($a['nullable']);
}

function proposed(array $a): int
{
    $nullable = $a['nullable'] ?? true;
    return $nullable === false ? 1 : 0;
}

$notDefined = [];
$nullable = ['nullable' => true];
$notNullable = ['nullable' => false];

echo PHP_EOL . 'not defined' . PHP_EOL;
var_dump(original($notDefined), proposed($notDefined));

echo PHP_EOL . 'nullable' . PHP_EOL;
var_dump(original($nullable), proposed($nullable));

echo PHP_EOL . 'not nullable' . PHP_EOL;
var_dump(original($notNullable), proposed($notNullable));
```

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6499Test.php [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    public $b;
+}
+
+/** @Entity */
+class DDC6499B
+{
+    /** @Id @Column(type="integer") @GeneratedValue */
+    public $id;
+
+    /** @ManyToOne(targetEntity="DDC6499A") */
```
[the remark this answers is not in the document]
Comment (bendavies) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6499Test.php]: fixed

File: tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6499Test.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (bendavies) [reply tests/Doctrine/Tests/ORM/Functional/Ticket/DDC6499Test.php]: fixed

File: lib/Doctrine/ORM/UnitOfWork.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (bendavies) [reply lib/Doctrine/ORM/UnitOfWork.php]: fixed

File: lib/Doctrine/ORM/UnitOfWork.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (bendavies) [reply lib/Doctrine/ORM/UnitOfWork.php]: btw, the original flawed (?) nullable check was introduced by @guilhermeblanco here: https://github.com/doctrine/doctrine2/pull/1570

**Comment by bendavies** [comment]
Ok, I've taken a deeper look at why `CascadeRemoveOrderTest` failing, and it is as follows:

1. `CascadeRemoveOrderEntityO` has a nullable FK to `CascadeRemoveOrderEntityG`
2. `CascadeRemoveOrderEntityG` has a nullable FK to `CascadeRemoveOrderEntityO`

This results in the `CommitOrderCalculator` having a node list of
`CascadeRemoveOrderEntityO -> CascadeRemoveOrderEntityG`, weight 0
`CascadeRemoveOrderEntityG -> CascadeRemoveOrderEntityO`, weight 0.

i.e this graph as 2 valid commit orders. It just so happens they with the fix in this PR, the "wrong" one is chosen, but we have no way of knowing it's the wrong one.

It seems that `CascadeRemoveOrderTest` only passes currently on 2.6 by depending on the bug that a non defined `nullable` results in weight 1.
Indeed, if you add `@JoinColumn(nullable=true)` to `CascadeRemoveOrderEntityG::$ownerO`, (essentially changing nothing), `CascadeRemoveOrderTest` will fail on 2.6/master.

So it seems we need to do something about the test. the fix in this PR seems valid.

Guidance appreciated @Ocramius @lcobucci @guilhermeblanco

**Comment by arnaud-lb** [comment]
I believe that the changes proposed in the PR are not enough to fix the problem, unfortunately.

With this PR applied, the code at https://gist.github.com/arnaud-lb/2ac4a3007bba08ca5f8fa78cfc353cf5 will generate the following graph:

```
<?php

$calc = new CommitOrderCalculator();

$calc->addNode('A', 'A');
$calc->addNode('B', 'B');
$calc->addNode('C', 'C');

// B referenced by A
$calc->addDependency('B', 'A', 1);

// A referenced by B (nullable)
$calc->addDependency('A', 'B', 0);

// A referenced by C
$calc->addDependency('A', 'C', 1);
```

The output of `$calc->sort()` is C, B, A. Since C has a non-nullable reference to A, trying to insert C will not work.

The problems seems to arise from the fact that we are trying to sort a cyclic graph topologically. When the execution reaches a state that shouldn't happen in the orignal algorithm (`case self::IN_PROGRESS`), the implementation doesn't abort, and nodes may be added to the sorted node list prematurely.

Ignoring the nullable references when building the graph seems to creates an acyclic graph in my case (Is a cycle of non-null references possible?), and the commit order is correct (on a quite complex schema). The order may not be optimal, though, and I don't know if it works for entity removal.

**Comment by bendavies** [comment]
thanks for that @arnaud-lb.
I'll add failing test cases to this PR.

**Comment by lcobucci** [comment]
@bendavies I think that we might consider to have a more complicated logic to calculate the weight of the dependencies. I mean, the `CommitOrderCalculator` is flexible enough to deal with other values (not only `0` and `1`) and we might be able to solve this issue by increasing the weight based on certain criteria, like:

```php
$weight = 0;

if ($joinColumnIsNullable) {
    ++$weight;
}

if ($isSomethingElse) {
    ++$weight;
}

// ...

return $weight;
```

**Comment by bendavies** [comment]
hah, i managed to hack in a fix for @arnaud-lb test

**Comment by bendavies** [comment]
@lcobucci I thinking that `CascadeRemoveOrderTest` is unsolvable by the `CommitOrderCalculator`.
`CascadeRemoveOrderEntityO` has a nullable FK to `CascadeRemoveOrderEntityG`
`CascadeRemoveOrderEntityG` has a nullable FK to `CascadeRemoveOrderEntityO`

O's FK could be populated, which means we need to delete O first, then G.
G's FK could be populated, which means we need to delete G first, then O.
Both O's and G's FK could be populated, in which case neither can be deleted first - It must be a 2 phase flush.

The only think i can think of (but i'm not very familiar with UOW) is to add `scheduleExtraUpdate` support for nullable foreign keys when removing entities, so you first null the FK, then remove the entity. I'm sure someone more experienced can tell me why that's a stupid idea.

**Comment by arnaud-lb** [comment]
> hah, i managed to hack in a fix for @arnaud-lb test/problem.

Awesome! Will it handle cycles involving more than 2 nodes, though?

**Comment by bendavies** [comment]
It's probably not robust at all. Like I said. A quick hack 😁.  can you
provide an example for a more complex graph?

**Comment by arnaud-lb** [comment]
Sure, here is a new failing test with a 3 nodes cycle: https://github.com/arnaud-lb/doctrine2/commit/c7facacefc3448f2af621c4c10579eb9829aca32

**Comment by bendavies** [comment]
Thanks @arnaud-ld. What is the correct commit order for that, to save me trying to work it out?

**Comment by arnaud-lb** [comment]
F, E, D, G would work as a commit order. Removing the addDependency() call when $joinColumnsNullable is true yields this order. However, I believe that doing this would cause more `scheduleExtraUpdate` than necessary with real world entities.

**Comment by Ocramius** [comment]
> @bendavies I think that we might consider to have a more complicated logic to calculate the weight of the dependencies. I mean, the `CommitOrderCalculator` is flexible enough to deal with other values (not only `0` and `1`)

@lcobucci that opens a can of worms IMO, as we start defining weights without having a clear idea of what the weights actually are.

 > It must be a 2 phase flush.

These are scenarios in which the ORM should probably kick in at metadata validation time... Would be interesting to see how far we can go with defining the commit order before runtime: possibly a completely different project?

 > EDIT: I've just seen that `CascadeRemoveOrderEntityO` has `onDelete="SET NULL"` on it's FK. Could the CommitOrderCalculator take that into account? However, if it didn't have `onDelete="SET NULL"`, the problem goes back to being unsolvable?

In short: no. Not all RDBMSs in use support `onDelete` or foreign keys in general. To take a simple one: MySQL's `onDelete` is already limited to 16 cascades, so it can't really be safely relied upon. We can't rely on that for now, sorry.

**Comment by bendavies** [comment]
@ocramius thanks for that. Do you have any opinion on `CascadeRemoveOrderTest` specifically? I still don't think the commit order calculator can determine the correct order for the classes in that test from class metadata alone, as it stands.

**Comment by Ocramius** [comment]
> Do you have any opinion on `CascadeRemoveOrderTest` specifically? 

It seems like the test was faulty due to missing explicit `nullable=true` there. Not much to do there: we simply unveiled a hidden issue here. I don't think it should be changed for now, as users might be affected.

**Comment by bendavies** [comment]
I've taken in @arnaud-lb failing tests and added the same scenario to the commit order calculator.
The graph is this:
image
It's obvious what the path should be from looking at it. The commit order calculator isn't currently clever enough to work this out.

**Comment by bendavies** [comment]
As above, at the moment the `CommitOrderCalculator` is too naive to work out some quite simple graphs.
I have a POC that fixes the above issues and improves the `CommitOrderCalculator` so that it solves the above graph and more complex ones.

Basically:
1. Find the [Minimum spanning tree](https://en.wikipedia.org/wiki/Minimum_spanning_tree) of the Graph.
2. Perform a [Topological Sort](https://en.wikipedia.org/wiki/Topological_sorting) of this graph to get your commit order.

However, there is a slight complication in that the `CommitOrderCalculator` could contain multiple unconnected graphs. MST requires a connected graph to perform on.
So before performing MST and sort, the distinct graphs need to be identified, and the above steps performed on each graph.

The sorts of these unconnected graphs can then be returned in any order as they are not dependent.

The above is a not a trivial undertaking so I don't want pursue it if someone thinks it's a bad idea or can see a problem I've overlooked.

If @guilhermeblanco could provide some feedback before I try and take this any further, that would be great, since he implemented the `CommitOrderCalculator` in it's current form.

**Comment by lcobucci** [comment]
> The commit order calculator isn't currently clever enough to work this out.

@bendavies @Ocramius the commit order calculator knows only about the `weight` it wasn't designed to know about the association mapping:

https://github.com/doctrine/doctrine2/blob/77e3e5c96c1beec7b28443c5b59145eeadbc0baf/lib/Doctrine/ORM/Internal/CommitOrderCalculator.php#L143-L148

It may indeed open a can of worms, but if we have issues with the logic which is sending data to the commit order calculator we shouldn't simply change the class but rather that logic.

**Comment by bendavies** [comment]
@lcobucci i think the data provided to the calculator is fine. This graph is simple but isn't solved correctly: https://github.com/doctrine/doctrine2/pull/7180#issuecomment-381341943

**Comment by arnaud-lb** [comment]
What's the status of this bug ? @bendavies you seemed to be on the right path.

**Comment by bendavies** [comment]
My solution wasn't good because `MST is for unidirected graphs while ours is directed.`

I was waiting for @guilhermeblanco's feedback/help, really.

**Comment by bendavies** [comment]
maybe fixed by https://github.com/doctrine/doctrine2/pull/7260

**Comment by bendavies** [comment]
rebased after #7260 was merged

**Comment by bendavies** [comment]
thought i'd take a look at how hibernate handles this. They do it a bit differently.
 
Doctrine deduces *table* insert ordering from class metadata, while hibernate orders the inserted entities themselves.

**Comment by ndench** [comment]
@bendavies how did you go with this?

**Comment by bendavies** [comment]
no further sorry.

**Comment by mpdude** [comment]
Still remember what you did 5 years ago? If so, maybe you can help to review #10547.
